include/nbl/video/ILogicalDevice.h

```diff
@@ -539,19 +543,133 @@ class ILogicalDevice : public core::IReferenceCounted
+            uploadedSize += subSize;
+        }
+        //return event.deferred_function; // ????? wtf is deferred_function? why would i have an event here??
+    }
+
+    inline void updateBufferRangeViaStagingBuffer(IGPUFence* fence, IGPUQueue* _queue, const asset::SBufferRange<IGPUBuffer>& bufferRange, const void* data)
+    {
+        core::smart_refctd_ptr<IGPUCommandPool> pool = createCommandPool(_queue->getFamilyIndex(), IGPUCommandPool::ECF_RESET_COMMAND_BUFFER_BIT);
+        core::smart_refctd_ptr<IGPUCommandBuffer> cmdbuf;
+        createCommandBuffers(pool.get(), IGPUCommandBuffer::EL_PRIMARY, 1u, &cmdbuf);
+        assert(cmdbuf);
+        cmdbuf->begin(IGPUCommandBuffer::EU_ONE_TIME_SUBMIT_BIT);
+        /*auto func = */updateBufferRangeViaStagingBuffer(cmdbuf.get(), fence, _queue, bufferRange, data);
+        cmdbuf->end();
+        IGPUQueue::SSubmitInfo submit;
+        submit.commandBufferCount = 1u;
+        auto* cmdbuf_raw = cmdbuf.get();
+        submit.commandBuffers = &cmdbuf_raw;
+        submit.signalSemaphoreCount = 0u;
+        submit.pSignalSemaphores = nullptr;
+        submit.waitSemaphoreCount = 0u;
+        submit.pWaitSemaphores = nullptr;
+        submit.pWaitDstStageMask = nullptr;
+        _queue->submit(1u, &submit, fence); // <threadsafeSubmit==true> means submit while locking an internal mutex
+        //func.optionalCmdBuffToDrop = std::move(cmdbuf); // ?????? wtf is optionalCmdBuffToDrop?
+    }
+
+    inline void updateBufferRangeViaStagingBuffer(IGPUQueue* _queue, const asset::SBufferRange<IGPUBuffer>& bufferRange, const void* data)
+    {
+        auto fence = this->createFence(static_cast<IGPUFence::E_CREATE_FLAGS>(0));
+        updateBufferRangeViaStagingBuffer(fence.get(), _queue, bufferRange, data);
+        auto* fenceptr = fence.get();
+        waitForFences(1u, &fenceptr, false, 9999999999ull);
+    }
+
     // Not implemented stuff:
     //vkCreateGraphicsPipelines //no graphics pipelines yet (just renderpass independent)
     //vkGetBufferMemoryRequirements
     //vkGetDescriptorSetLayoutSupport
-    //vkMapMemory
-    //vkUnmapMemory
     //vkTrimCommandPool
     //vkGetPipelineCacheData //as pipeline cache method??
     //vkMergePipelineCaches //as pipeline cache method
     //vkCreateQueryPool //????
     //vkCreateShaderModule //????
 
 protected:
+    // must be called by implementations of mapMemory()
+    static void post_mapMemory(IDriverMemoryAllocation* memory, void* ptr, IDriverMemoryAllocation::MemoryRange rng, IDriverMemoryAllocation::E_MAPPING_CPU_ACCESS_FLAG access) 
+    {
+        memory->postMapSetMembers(ptr, rng, access);
+    }
+    // must be called by implementations of unmapMemory()
+    static void post_unmapMemory(IDriverMemoryAllocation* memory)
+    {
+        post_mapMemory(memory, nullptr, { 0,0 }, IDriverMemoryAllocation::EMCAF_NO_MAPPING_ACCESS);
+    }
```

Comment:
this is a mess, `ILogicalDevice` is already a friend, so get rid of the `postMapSetMembers` function from the `IDriverMemoryAllocation` and do this stuff directly in `ILogicalDevice`

Reply:
idk, seems like a greater mess then

Reply:
no this is certainly worse, I have 3 functions which are badlynamed reference each other, all to do what is basically 3 lines of code each